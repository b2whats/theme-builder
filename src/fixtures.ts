import { Tokens } from './Tokens'
import { Properties } from './Properties'
import { Component } from './Component'

export const tokens = new Tokens({
  focus: '0px 0px 2px 3px black',
  breakpoints: ['70em', '50em', '30em'],
  shadow: {
    1: '0px 1px 2px black',
    2: '3px 4px 5px black',
  },
  font: {
    fontSize: {
      l: 15,
      m: 13,
      s: 11,
    },
  },
  palette: {
    red:  '#FFEDEE',
    blue: '#FFE6E6',
    green: '#FFC5C6',
    primary: 'blue'
  },
})


export const properties = new Properties(tokens)
  .breakpoints({
    token: 'breakpoints'
  })
  .add('display', {
    toString: (value: 'block' | 'flex' | 'inline') => `display: ${value};`
  })
  .add('color', {
    token: 'palette', 
    toString: (value: string) => `color: ${value};`
  })
  .add('fontSize', {
    token: 'font.fontSize',
    toString: (value) => `font-size: ${value}px;`
  })
  .add('shadow', {
    token: 'shadow',
    toString: (value) => `box-shadow: ${value};`
  })
  .add('focus', {
    token: 'focus',
    toString: (value) =>  value ? `
      &&:focus{
        box-shadow: ${value};
        position: relative;
        z-index: 2;
      }
    ` : ''
  })
  .states({
    hover: (rules) => `&&:hover{${rules}}`
  })
  //.list()

type Props = {
  /** Размер */
  size: 's' | 'm' | 'l'
  kind: 'fill' | 'outline'
}

export const component = new Component(properties)
  .name('Button')
  .types<Props>()
  .defaultProps({
    size: 's'
  })
  .slot('base', {
  })
  .slot('base2', {
    display: 'block',
    hover: {
      color: ['primary']
    }
  })

