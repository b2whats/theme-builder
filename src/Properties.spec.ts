import { properties, tokens } from './fixtures'

test('Properties compute - without token', () => {
  const rule = properties.compute('display', 'block', tokens.scheme)
  const cache = properties.hasCache('display', 'block', tokens.scheme)

  expect(cache).toBe(true)
  expect(rule).toMatchSnapshot()
})

test('Properties compute - with token resolve', () => {
  const rule = properties.compute('color', 'red', tokens.scheme)
  const cache = properties.hasCache('color', 'red', tokens.scheme)

  expect(cache).toBe(true)
  expect(rule).toMatchSnapshot()
})

test('Properties compute - with token no resolve', () => {
  const rule = properties.compute('color', '#hex', tokens.scheme)
  const cache = properties.hasCache('color', '#hex', tokens.scheme)

  expect(cache).toBe(true)
  expect(rule).toMatchSnapshot()
})

test('Properties compute - with token resolve (deep)', () => {
  const rule = properties.compute('fontSize', 's', tokens.scheme)
  const cache = properties.hasCache('fontSize', 's', tokens.scheme)

  expect(cache).toBe(true)
  expect(rule).toMatchSnapshot()
})

test('Properties compute - with token(number) resolve', () => {
  const rule = properties.compute('shadow', 1, tokens.scheme)
  const cache = properties.hasCache('shadow', 1, tokens.scheme)

  expect(cache).toBe(true)
  expect(rule).toMatchSnapshot()
})

test('Properties compute - with token(boolean) resolve', () => {
  const rule = properties.compute('focus', true, tokens.scheme)
  const cache = properties.hasCache('focus', true, tokens.scheme)

  expect(cache).toBe(true)
  expect(rule).toMatchSnapshot()
})

test('Properties compute - with token(boolean) no resolve', () => {
  const rule = properties.compute('focus', false, tokens.scheme)
  const cache = properties.hasCache('focus', false, tokens.scheme)

  expect(cache).toBe(true)
  expect(rule).toMatchSnapshot()
})

test('Properties compute - with token(boolean array) resolve', () => {
  const rule = properties.compute('focus', [true, false], tokens.scheme)
  const cache = properties.hasCache('focus', [true, false], tokens.scheme)

  expect(cache).toBe(true)
  expect(rule).toMatchSnapshot()
})

test('Properties compute - with token resolve (deep array)', () => {
  const rule = properties.compute('fontSize', ['s', 'm', 'l'], tokens.scheme)
  const cache = properties.hasCache('fontSize', ['s', 'm', 'l'], tokens.scheme)

  expect(cache).toBe(true)
  expect(rule).toMatchSnapshot()
})

test('Properties compute - with token resolve (deep array nullish)', () => {
  const rule = properties.compute('fontSize', ['s', , 'l'], tokens.scheme)
  const cache = properties.hasCache('fontSize', ['s', , 'l'], tokens.scheme)

  expect(cache).toBe(true)
  expect(rule).toMatchSnapshot()
})