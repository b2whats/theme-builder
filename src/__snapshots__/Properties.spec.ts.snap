// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Properties compute - with token no resolve 1`] = `"color: #hex;"`;

exports[`Properties compute - with token resolve (deep array nullish) 1`] = `
Array [
  "font-size: 11px;",
  undefined,
  "font-size: 15px;",
]
`;

exports[`Properties compute - with token resolve (deep array) 1`] = `
Array [
  "font-size: 11px;",
  "font-size: 13px;",
  "font-size: 15px;",
]
`;

exports[`Properties compute - with token resolve (deep) 1`] = `"font-size: 11px;"`;

exports[`Properties compute - with token resolve 1`] = `"color: #FFEDEE;"`;

exports[`Properties compute - with token(boolean array) resolve 1`] = `
Array [
  "
      &&:focus{
        box-shadow: 0px 0px 2px 3px black;
        position: relative;
        z-index: 2;
      }
    ",
  "",
]
`;

exports[`Properties compute - with token(boolean) no resolve 1`] = `""`;

exports[`Properties compute - with token(boolean) resolve 1`] = `
"
      &&:focus{
        box-shadow: 0px 0px 2px 3px black;
        position: relative;
        z-index: 2;
      }
    "
`;

exports[`Properties compute - with token(number) resolve 1`] = `"box-shadow: 0px 1px 2px black;"`;

exports[`Properties compute - without token 1`] = `"display: block;"`;
